components/tel/Core/Src/freertos.c

```diff
@@ -0,0 +1,462 @@
+
+  /* USER CODE BEGIN RTOS_QUEUES */
+
+  canMessageQueueHandle = osMessageQueueNew(CAN_MESSAGE_QUEUE_SIZE, sizeof(CAN_msg_t), &canMessageQueue_attributes);
+  imuMessageQueueHandle = osMessageQueueNew(IMU_QUEUE_SIZE, sizeof(IMU_msg_t), &imuMessageQueue_attributes);
+  /* USER CODE END RTOS_QUEUES */
+
+  /* Create the thread(s) */
+  /* creation of defaultTask */
+  // defaultTaskHandle = osThreadNew(StartDefaultTask, NULL, &defaultTask_attributes);
+
+  /* USER CODE BEGIN RTOS_THREADS */
+
+  // kernelLEDTaskHandle = osThreadNew(kernelLEDTask, NULL, &kernelLEDTask_attributes);
+  
+  readCANTaskHandle = osThreadNew(readCANTask, NULL, &readCANTask_attributes);
+  transmitMessageTaskHandle = osThreadNew(transmitMessageTask, NULL, &transmitMessageTask_attributes);
+
+  readimuHandle = osThreadNew(StartReadIMU, NULL, &readimu_attributes);
+  transmitIMUHandle = osThreadNew(StartTransmitIMU, NULL, &transmitIMU_attributes);
+
+  // changeRadioSettingsTaskHandle = osThreadNew(changeRadioSettingsTask, NULL, &changeRadioSettingsTask_attributes);
+
+  /* USER CODE END RTOS_THREADS */
+
+  /* USER CODE BEGIN RTOS_EVENTS */
+  /* add events, ... */
+  /* USER CODE END RTOS_EVENTS */
+
+}
+
+/* USER CODE BEGIN Header_StartDefaultTask */
+/**
+  * @brief  Function implementing the defaultTask thread.
+  * @param  argument: Not used
+  * @retval None
+  */
+/* USER CODE END Header_StartDefaultTask */
+void StartDefaultTask(void *argument)
+{
+  /* USER CODE BEGIN StartDefaultTask */
+  /* Infinite loop */
+  for(;;)
+  {
+    osDelay(1);
+  }
+  /* USER CODE END StartDefaultTask */
+}
+
+/* Private application code --------------------------------------------------*/
+/* USER CODE BEGIN Application */
+
+__NO_RETURN void kernelLEDTask (void *argument) {
+  osKernelState_t kernel_status;
+
+  while (1) {
+    kernel_status = osKernelGetState();
+
+    if (kernel_status == osKernelRunning) {
+      HAL_GPIO_TogglePin(KERNEL_LED_GPIO_Port, KERNEL_LED_Pin);
+    }
+
+    osDelay(KERNEL_LED_DELAY);
+  }
+}
+
+__NO_RETURN void readCANTask(void *argument) {
+  static HAL_StatusTypeDef rx_status;
+  static CAN_msg_t current_can_message;
+
+  while (1) {
+    // wait for thread flags to be set
+    osThreadFlagsWait(CAN_READY, osFlagsWaitAll, osWaitForever);
+
+    if (HAL_CAN_GetRxFifoFillLevel(&hcan, CAN_RX_FIFO0) != 0) {
+      // there are multiple CAN IDs being passed through the filter, pull out the current message
+      rx_status = HAL_CAN_GetRxMessage(&hcan, CAN_RX_FIFO0, &can_rx_header, current_can_data);
+      HAL_GPIO_TogglePin(KERNEL_LED_GPIO_Port, KERNEL_LED_Pin);
+
+      // package into CAN_msg_t
+      current_can_message.header = can_rx_header;
+
+
+      for (uint8_t i=0; i<8; i++) {
+        current_can_message.data[i] = current_can_data[i];
+      }
+
+      // priority=0, timeout=0
+      osMessageQueuePut(canMessageQueueHandle, &current_can_message, 0U, 0U);
+    }
+
+    HAL_CAN_ActivateNotification(&hcan, CAN_IT_RX_FIFO0_MSG_PENDING);
```

Comment:
If you go to the HAL documentation you'll see a section with steps to use CAN rx with interrupts. It may be helpful to comment here that this allows the next interrupt to trigger again.